### ISSUE TITLE
big endian test failures s390x architecture

### ISSUE THREAD

**Issue by emollier** [biopython/biopython#4658]
### Setup

I am reporting a problem with Biopython version, Python version,
and operating system as follows.

	>>> import sys; print(sys.version)                       
	3.12.2 (main, Mar  3 2024, 09:11:00) [GCC 13.2.0]
	>>> import platform; print(platform.python_implementation()); print(platform.platform())
	CPython
	Linux-6.7.9-amd64-s390x-with-glibc2.37
	>>> import Bio; print(Bio.__version__)
	1.83

### Expected behaviour

I would like to make all test items of the test suite pass on
Debian sid for s390x CPU architecture, or justify appropriately
their skipping.  Note that s390x is a big endian architecture.

### Actual behaviour

The following tests are failing:

  * test_Align_bigbed.TestAlign_extended_bed.test_writing
  * test_Tutorial.TutorialTestCase.test_doctests

test_Align_bigbed.TestAlign_extended_bed.test_writing has for
relevant error message (which very much looks like bytes have
been swabbed, it might be that reference data needs adjustment
on big endian platforms, or the function being tested needs to
take endianness into account):

	Traceback (most recent call last):
	 File "/<<PKGBUILDDIR>>/.pybuild/cpython3_3.12/build/Tests/test_Align_bigbed.py", line 2503, in test_writing
	   self.assertEqual(correct, data)
	AssertionError: b'\xeb\xf2\x89\x87\x04\x00\x03\x00a\x04\x00\x[153321 chars]\x87' != b'\x87\x89\xf2\xeb\x00\x04\x00\x03\x00\x00\x0[153323 chars]\xeb'

test_Tutorial.TutorialTestCase.test_doctests has for relevant
error message:

	Traceback (most recent call last):
	 File "/<<PKGBUILDDIR>>/.pybuild/cpython3_3.12/build/Tests/test_Tutorial.py", line 244, in test_doctests
	   raise ValueError(
	ValueError: 1 Tutorial doctests failed: test_chapter_align_line_00019

You can find a build log complete with full test output on the
Debian buildd [build log for python-biopython for s390x].
Please disregard all the noise produced by test_GenomeDiagram
errors.  They seem related to a failure to fetch the font file
/usr/share/fonts/type1/urw-base35/C059-Roman.PFB, which looks to
be symptomatic of a transient problem in the testbed at the
moment: I reproduced the problem on today's Debian sid amd64.

[build log for python-biopython for s390x]: https://buildd.debian.org/status/fetch.php?pkg=python-biopython&arch=s390x&ver=1.83%2Bdfsg-1&stamp=1710314200&raw=1

### Steps to reproduce

Run the test suite on Debian sid on s390x platform, qemu is an
option.

---

I am not in a rush, as I am not expecting that many users of
Biopython on s390x, but I thought you might be interested of the
issues anyway.  We (Debian) are in the middle of the 64-bit
time_t transition: one of the largest bulk change in the
operating system, possibly since the libc5 to libc6 transition.
I'll probably consider skipping the tests if properly patching
were to take time.

Have a nice day,  :)
Étienne.

**Comment by mdehoon** [comment]
Thanks for the report. It seems like the bug is in the test itself.
I don't have access to a big-endian machine. Would you be able to generate the `bigbed_extended.bb` file using UCSC's `bedToBigBed` utility on s390x? Then Biopython can include a big-endian and a little-endian version of this file, which should fix the problem.

**Comment by emollier** [comment]
I eventually found out the [bedToBigBed source code] to recompile against s390x and biopython's instructions:
```
    # The bigBed file bigbed_extended.bb is a BED9+2 file, with nine predefined
    # BED fields and 2 extra (custom) fields. It was created by running
    #
    # bedToBigBed -as=bedExample2.as -type=bed9+2 -extraIndex=name,geneSymbol bedExample2.bed hg18.chrom.sizes bigbed_extended.bb
    #
    # where bedExample2.bed contains 10 lines selected from the example BED9+2
    # file bedExample2.bed downloaded from UCSC
    # (https://genome.ucsc.edu/goldenPath/help/examples/bedExample2.bed)
    # and bedExample2.as the associated AutoSQL file, also downloaded from UCSC
    # (https://genome.ucsc.edu/goldenPath/help/examples/bedExample2.as)
    # declaring the nine predefined BED fields and the two extra fields
```
I think I probably missed one of the instructions (that could be the 10 lines selection perhaps, which I haven't understood, but I also failed to locate an hg18.chrom.sizes file and used biopython's hg38.chrom.sizes instead) because my resulting file is one order of magnitude larger than the little-endian equivalent:
```
$ du -sh bigbed_extended*
40K     bigbed_extended.bb
360K    bigbed_extended_be.bb
```
How should I filter the bedExample2.bed file to get the appropriate lines, if picking everything went wrong?  Do you have an idea where I could find the hg18.chrom.sizes, if this is not a typo?

Thanks,
Étienne.

[bedToBigBed source code]: https://github.com/ucscGenomeBrowser/kent/blob/master/src/utils/bedToBigBed/bedToBigBed.c

**Comment by mdehoon** [comment]
Thank you. You can generate the bedExample2.bed file from the bigbed_extended.bb in the Biopython source distribution (in the Tests/Blat subdirectory):

bigBedToBed Tests/Blat/bigbed_extended.bb bedExample2.bed

This will generate the 10-line bedExample2.bed file.

The bedExample2.as is in the same subdirectory Tests/Blat .

You can get hg18.chrom.sizes from UCSC at http://hgdownload.cse.ucsc.edu/goldenpath/hg18/bigZips/hg18.chrom.sizes .

**Comment by emollier** [comment]
Thanks, it took me some struggle juggling with the different archs and files erased, but I managed to get somewhere.  From visual inspection with an hexadecimal editor, the file below looks to be the big endian equivalent (once gunzipped) of bigbed_extended.bb.

[bigbed_extended_be.bb.gz](https://github.com/biopython/biopython/files/14628580/bigbed_extended_be.bb.gz)

**Comment by mdehoon** [comment]
@emollier  Thank you. Can you try to see if all tests pass after #4670 ?

**Comment by emollier** [comment]
I confirm test_Align_bigbed.TestAlign_extended_bed.test_writing is resoved on s390x without introducing regressions on amd64.  Thank you!  :)

The test_Tutorial.TutorialTestCase.test_doctests failure is still around though:
```
======================================================================
ERROR: test_doctests (test_Tutorial.TutorialTestCase.test_doctests)
Run tutorial doctests.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/.pybuild/cpython3_3.11/build/Tests/test_Tutorial.py", line 244, in test_doctests
    raise ValueError(
ValueError: 1 Tutorial doctests failed: test_chapter_align_line_00019
```

(I've also been reminded to have a look at the missing font issue, but that would be another item for another issue, perhaps not even against biopython.)

**Comment by peterjc** [comment]
For more details on the failing Tutorial example, can you try:

```
$ cd Tests
$ python test_Tutorial.py
Running Tutorial doctests...
Tests done
```

That will be more verbose. Thanks!

**Comment by emollier** [comment]
Thanks for the hint, I could determine the issue stemmed from an endian dependent output in the doctest.  I could work it around with the following patch, and this resolved the test failure, but only on s390x since I broke tests on little endian architectures:
```patch
--- python-biopython.orig/Doc/Tutorial/chapter_align.tex
+++ python-biopython/Doc/Tutorial/chapter_align.tex
@@ -483,7 +483,7 @@
 >>> align_array  # doctest: +NORMALIZE_WHITESPACE
 array([['C', 'G', 'G', 'T', 'T', 'T', 'T', 'T'],
        ['A', 'G', '-', 'T', 'T', 'T', '-', '-'],
-       ['A', 'G', 'G', 'T', 'T', 'T', '-', '-']], dtype='<U1')
+       ['A', 'G', 'G', 'T', 'T', 'T', '-', '-']], dtype='>U1')
 \end{minted}
 Note that the \verb+alignment+ object and the NumPy array \verb|align_array| are separate objects in memory - editing one will not update the other!
 
```
I'm concerned that providing a proper fix may clutter the documentation, so I'm happy to skip this test item on big endian architectures.  If that works for you, then I think we can consider the present case closed.

By the way, the font issue affecting test_GenomeDiagram is now gone.  I assume something went off with the dependency, then promptly fixed.  So everything is good from my perspective.

Thank you both for your help!
Have a nice day,  :)
Étienne.

**Comment by peterjc** [comment]
Nice work! Yes, now that you've found the error it is very clear cut - the NumPy array's type is coming through as big or little endian: https://numpy.org/doc/stable/reference/generated/numpy.dtype.byteorder.html

@mdehoon may be able to see a sensible code change, e.g. is it better to get it to use "=" which means native?

Or, we modify the example - probably just to skip testing that particular output:

https://github.com/biopython/biopython/blob/biopython-183/Doc/Tutorial/chapter_align.tex#L466

(I'm loath to further complicate the Tutorial testing with big/little Endian switching)

**Comment by mdehoon** [comment]
>may be able to see a sensible code change, e.g. is it better to get it to use "=" which means native?

"=" can be used when creating the array, but when printing the array, the dtype will still be "<..." or ">..." depending on whether the system is little- or big-endian.

**Comment by peterjc** [comment]
I think we're OK to close this issue now on the Biopython side, do you agree @emollier?

Also thank you again (and others like you doing packaging). Over the years we've had many useful bug reports like this with followup investigation on a platform or system we don't have easy access too. This is appreciated.

**Comment by emollier** [comment]
All good for me, I'm closing the issue.
Have a nice day,  :)
Étienne.
